### ISSUE TITLE
Future plans question

### ISSUE THREAD

**Issue by keithZmudzinski** [antmicro/verilator-dynamic-scheduler-examples#4]
I'm interested in using this project, it seems to be exactly what I need :)
Are there plans to have this merged into the official Verilator repo?

**Comment by tgorochowik** [comment]
Yes, most certainly!

This repository shows that the dynamic scheduler works and we would very much like to include that (as an option) in official Verilator. Before we can do that some more work is needed first - for now we completely replace the stock static scheduler - the dynamic scheduler allows us to implement much more features (which you can see in this repo) but it is not necessarily better in other aspects (e.g. speed of execution) - therefore making it optional is a must.

**Comment by Blebowski** [comment]
Hello,
may I ask what is the current status of this work? I see many different branches in: https://github.com/antmicro/verilator, and also in: https://github.com/antmicro/verilator-1. When could we expect dynamic scheduling being merged into verilator mainline ?

**Comment by kbieganski** [comment]
Hi,
The current status is being tracked [here](https://antmicro.github.io/verilator-dynamic-scheduler-tests/report.html). It shows how many regression tests are passing with the dynamic scheduler. Once we get it to 100% (and introduce a switch to optionally enable it), we'll open a PR to mainline. That should be early 2022. It's harder to predict when it will get merged, but we'll try to do it as quickly as we can!

As for where you can find it, [here](https://github.com/antmicro/verilator-1/tree/dynamic-scheduler)'s the active branch with the newest tested revision.

**Comment by Blebowski** [comment]
Hi @kbieganski, thanks for quick reply. I am looking forward for this feature. I totally understand that merging can take lot of time since amount of changes is huge. 

Btw, does that mean that it will be already possible to compile UVM testbench with Verilator? Or is it still long road ahead?

**Comment by kbieganski** [comment]
Unfortunately, there's still a bunch of other features needed for UVM, including the `process` class, randomize constraints, better assertion support, probably more I'm forgetting right now. And even the dynamic scheduler will surely need some maturing. Hopefully having it in mainline will accelerate work on the other stuff, whether it's from us or other people in the community.

**Comment by Blebowski** [comment]
Ok, I see, thanks for explaining. However, with this change it should be possible to implement e.g. simple clock generator with blocking assigns and #(half_period), within the Verilated code, right ? Or is there still need to create C/C++ wrapper around verilated code?

**Comment by kbieganski** [comment]
Yes, it would work, and in fact that is one of the examples given in this repo.

**Comment by kbieganski** [comment]
@keithZmudzinski @Blebowski In case you're still interested, dynamic scheduling (or timing, as it's now being referred to) has landed in the Verilator 5 testing branch: https://github.com/verilator/verilator/tree/develop-v5
Check it out! Testers are welcome :)

PS. You can avoid writing the C++ wrappper by passing `--main` to Verilator. Together with `--timing` this means you don't need to write any C++ code.

**Comment by raulbehl** [comment]
@kbieganski Congratulations on achieving this milestone. I'd be up for testing it over the weekend.
